### PR TITLE
[TechDraw] Improve UI for TaskSurfaceFinishSymbols

### PR DESCRIPTION
**Pull request by benj5378**
Screenshot from 2024-08-29 23-43-36
Before

Screenshot from 2024-08-30 00-34-55
After

Who don't like better UI? #12719

### CODE REVIEW

**Comment by maxwxyz** [comment]
@FreeCAD/design-working-group FYI

**Comment by pierreporte** [comment]
The proposal looks good.

**Comment by Syres916** [comment]
Thanks for the efforts, my eye strain will be reduced significantly with this improvement.

**Comment by obelisk79** [comment]
Looks good

File: src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp [hunk 1]
```diff
@@ -336,11 +361,14 @@ void TaskSurfaceFinishSymbols::onIconChanged()
...
     QIcon symbolIcon = pressedButton->icon();
-    QGraphicsPixmapItem* pixmapItem = new(QGraphicsPixmapItem);
-    pixmapItem->setPixmap(symbolIcon.pixmap(50, 64));
-    pixmapItem->setPos(-50, -130);
-    pixmapItem->setZValue(-1);
-    symbolScene->addItem(pixmapItem);
+    if(currentIcon) {
+        symbolScene->removeItem(currentIcon);
+    }
+    currentIcon = new(QGraphicsPixmapItem);
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp:367]: Hmmm, would this cause a memory leak?

Comment (kadet1090) [reply src/Mod/TechDraw/Gui/TaskSurfaceFinishSymbols.cpp:367]: If it does not have parent set properly it can.

**Comment by Syres916** [comment]
@benj5378 based on https://stackoverflow.com/questions/42702280/how-to-update-qgraphicsscene-in-a-qmainwindow do you actually have to keep removing and then adding an item to the scene when the user clicks on a different button, can you not just apply the different pixmap which will get repainted hence reducing the chance of a segfault?

**Comment by benj5378** [comment]
> @benj5378 based on https://stackoverflow.com/questions/42702280/how-to-update-qgraphicsscene-in-a-qmainwindow do you actually have to keep removing and then adding an item to the scene when the user clicks on a different button, can you not just apply the different pixmap which will get repainted

No, I don't have to. I could just update the pixmap

> hence reducing the chance of a segfault?

I don't see how it reduces the chance of a segfault?

**Comment by Syres916** [comment]
> I don't see how it reduces the chance of a segfault?

I'll try to write a reliable Gui Unit Test to hopefully demonstrate what I found in a very short time of testing but here's the manual version:

The basic steps are to rename your existing config files so you are starting as a new user.
On starting FreeCAD only change the Nav Style to what you prefer, do not click on any theme and then click Done to close the First time Start widget.
Click to create an Empty File
Activate the overlay on the Model Dock Widget and set Transparency on
Set Tasks to a floating window and move to the right hand side, dock it and set Transparency on
Close FreeCAD to write all the basic parameters to the config files
Start FreeCAD and open an existing file that contains a Techdraw Projection Group
Double-click the Projection Group to open it in Tasks and click Cancel
Start the Surface Finish Symbols command and without moving the mouse or clicking anything I get:

```
Program received signal SIGSEGV, Segmentation fault.
#0  /lib/x86_64-linux-gnu/libc.so.6(+0x45320) [0x734008045320]
#1  0x734009a924d8 in QGraphicsItem::scene() const from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x8
#2  0x734009ac9c8f in QGraphicsScene::removeItem(QGraphicsItem*) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x3f
#3  0x733fbaeda6d4 in TechDrawGui::TaskSurfaceFinishSymbols::onIconChanged() from /home/username/freecad-daily-build/Mod/TechDraw/TechDrawGui.so+0xa4
#4  /lib/x86_64-linux-gnu/libQt5Core.so.5(+0x312e16) [0x734008b12e16]
#5  0x7340098676e4 in QAbstractButton::clicked(bool) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x44
#6  /lib/x86_64-linux-gnu/libQt5Widgets.so.5(+0x26796f) [0x73400986796f]
#7  0x7340098699a6 in QAbstractButton::click() from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x96
#8  0x733fbaed95c7 in TechDrawGui::TaskSurfaceFinishSymbols::TaskSurfaceFinishSymbols(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) from /home/username/freecad-daily-build/Mod/TechDraw/TechDrawGui.so+0xa57
#9  0x733fbaeda8d1 in TechDrawGui::TaskDlgSurfaceFinishSymbols::TaskDlgSurfaceFinishSymbols(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) from /home/username/freecad-daily-build/Mod/TechDraw/TechDrawGui.so+0x51
#10  0x733fbadf3146 in CmdTechDrawSurfaceFinishSymbols::activated(int) from /home/username/freecad-daily-build/Mod/TechDraw/TechDrawGui.so+0x1c6
#11  0x73400bb07996 in Gui::Command::_invoke(int, bool) from /home/username/freecad-daily-build/lib/libFreeCADGui.so+0x226
#12  0x73400bb07d76 in Gui::Command::invoke(int, Gui::Command::TriggerSource) from /home/username/freecad-daily-build/lib/libFreeCADGui.so+0x136
#13  /lib/x86_64-linux-gnu/libQt5Core.so.5(+0x312e16) [0x734008b12e16]
#14  0x734009764f94 in QAction::triggered(bool) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x44
#15  0x734009767eab in QAction::activate(QAction::ActionEvent) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0xbb
#16  /lib/x86_64-linux-gnu/libQt5Widgets.so.5(+0x2fc512) [0x7340098fc512]
#17  /lib/x86_64-linux-gnu/libQt5Widgets.so.5(+0x304702) [0x734009904702]
#18  0x7340097b0df8 in QWidget::event(QEvent*) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x278
#19  0x73400976bd45 in QApplicationPrivate::notify_helper(QObject*, QEvent*) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x85
#20  0x7340097746b0 in QApplication::notify(QObject*, QEvent*) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x1290
#21  0x73400baa4a38 in Gui::GUIApplication::notify(QObject*, QEvent*) from /home/username/freecad-daily-build/lib/libFreeCADGui.so+0x98
#22  0x734008ad8118 in QCoreApplication::notifyInternal2(QObject*, QEvent*) from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x128
#23  0x734009772874 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool, bool) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x1d4
#24  /lib/x86_64-linux-gnu/libQt5Widgets.so.5(+0x1cb5d2) [0x7340097cb5d2]
#25  /lib/x86_64-linux-gnu/libQt5Widgets.so.5(+0x1cdfbf) [0x7340097cdfbf]
#26  0x73400976bd45 in QApplicationPrivate::notify_helper(QObject*, QEvent*) from /lib/x86_64-linux-gnu/libQt5Widgets.so.5+0x85
#27  0x73400baa4a38 in Gui::GUIApplication::notify(QObject*, QEvent*) from /home/username/freecad-daily-build/lib/libFreeCADGui.so+0x98
#28  0x734008ad8118 in QCoreApplication::notifyInternal2(QObject*, QEvent*) from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x128
#29  0x734008f45a3b in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) from /lib/x86_64-linux-gnu/libQt5Gui.so.5+0x80b
#30  0x734008f17bfc in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) from /lib/x86_64-linux-gnu/libQt5Gui.so.5+0xac
#31  /lib/x86_64-linux-gnu/libQt5XcbQpa.so.5(+0x75d06) [0x734002848d06]
#32  /lib/x86_64-linux-gnu/libglib-2.0.so.0(+0x5d5b5) [0x734006d145b5]
#33  /lib/x86_64-linux-gnu/libglib-2.0.so.0(+0xbc717) [0x734006d73717]
#34  /lib/x86_64-linux-gnu/libglib-2.0.so.0(g_main_context_iteration+0x33) [0x734006d13a53]
#35  0x734008b35279 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x69
#36  0x734008ad6a7b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x13b
#37  0x734008adf3e8 in QCoreApplication::exec() from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x98
#38  0x73400b9c8449 in Gui::Application::runApplication() from /home/username/freecad-daily-build/lib/libFreeCADGui.so+0x9d9
#39  /home/username/freecad-daily-build/bin/FreeCAD(+0x828c) [0x5a241405528c]
#40  /lib/x86_64-linux-gnu/libc.so.6(+0x2a1ca) [0x73400802a1ca]
#41  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x8b) [0x73400802a28b]
#42  /home/username/freecad-daily-build/bin/FreeCAD(+0x8665) [0x5a2414055665]

```
It's not 100% reproduce able but way too frequent for me to actually consider using, I'll put up with the eye strain instead.

When it does work I get annoying warnings such as:

`QGraphicsScene::removeItem: item 0x5d50cddd4c60's scene (0x5d50cddd6d10) is different from this scene (0x5d50d073b8c0)`

```
OS: Linux Mint 22 (X-Cinnamon/cinnamon)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38577 (Git)
Build type: Release
Branch: main
Hash: 2a9cf9e8e6e62086948eb15e24d1e69a54f53d97
Python 3.12.3, Qt 5.15.13, Coin 4.0.2, Vtk 9.1.0, OCC 7.6.3
Locale: English/United Kingdom (en_GB)
Installed mods: 
  * FreeCAD-themes 2024.7.24
  * Silk 0.1.5
  * sheetmetal 0.4.24
  * Behave-Dark-Colors 0.1.1
  * CfdOF 1.27.5
  * freecad.gears 1.2.0
  * Curves 0.6.45
  * A2plus 0.4.66
```

So from my personal point of view the adding and removing of the item if it's not entirely necessary (which I don't think it is) should be avoided.

I have a Qt6.7.2 EndeavorOS vm, a Debian 12 with Qt 5.15.8 vm and an older Linux Mint 21.3 vm with Qt5.15.3 which I can carry out the test later on as well.

**Comment by benj5378** [comment]
@Syres916 I think I figured out, what was going on here. `currentIcon` was never intialized meaning, the value was whatever is in the memory. Sometimes, there was a `nullptr` creating the segfault. Other times, there was a previous `QGraphicsItem*` - then trying remove a random `QGraphicsItem` in memory from the scene.

I now made sure that `currentIcon` is `initialized` to nullptr. Can you check, if you are still able to reproduce this issue? If yes, please give (new) backtraces

**Comment by Syres916** [comment]
OK, that last change certainly has improved stability, I've not received a seg fault since, thanks.

**Comment by benj5378** [comment]
Fixed #16272

**Pull request merged**